scrapy/utils/datatypes.py

```diff
@@ -304,3 +305,54 @@ def __setitem__(self, key, value):
         while len(self) >= self.limit:
             self.popitem(last=False)
         super(LocalCache, self).__setitem__(key, value)
+
+
+class ExpiringCache(LocalCache):
+    """
+
+    Dictionary with finite number of keys and time-based expiration.
+
+    Membership tests (key in dict, dict.has_key) are not guaranteed to trigger
+    expiration.
+
+    """
+    def __init__(self, limit=None, expiration=None):
+        super(ExpiringCache, self).__init__(limit)
+        self.expiration = expiration
+
+    def __getitem__(self, key):
+        setting_time, value = super(ExpiringCache, self).__getitem__(key)
+        if self.expiration and setting_time + self.expiration < time.time():
+            del self[key]
+            raise KeyError("key expired")
+        return value
+
+    def __setitem__(self, key, value):
+        super(ExpiringCache, self).__setitem__(key, (time.time(), value))
+
+    def get(self, key, default=None):
+        try:
+            return self[key]
+        except KeyError:
+            return default
+
+    def iteritems(self):
+        for k in self.iterkeys():
+            yield k, self[k]
+
+    def itervalues(self):
+        for k in self.iterkeys():
+            yield self[k]
+
+    if six.PY2:
+        def items(self):
+            return list(self.iteritems())
+
+        def values(self):
+            return list(self.itervalues())
+    else:
+        def iterkeys(self):
+            # For iteritems() and itervalues() above.
+            return self.keys()
+        items = iteritems
+        values = itervalues
```

Comment:
`items()` and `values()` are inherited from OrderedDict. Starting with Python 3.5, they will bypass the custom `__getitem__` (because OrderedDict is implemented in C now) and expose the actual `(time, value)` tuple.

In Python up to 3.4:

    >>> list(d.items())
    [('a', 1)]

In Python 3.5:

    >>> list(d.items())
    [('a', (1481032887.784956, 1))]